k2/csrc/array.h

```diff
@@ -173,7 +174,24 @@ class Array1 {
   // Convert this array to another array with type S;
   // if S is same with T, then it just returns *this
   template <typename S>
-  Array1<S> AsType();
+  Array1<typename std::enable_if<!std::is_same<S, T>::value, S>::type> AsType()
```

Comment:
would be better in array_inl.h IMO, but OK.

Reply:
Will get `function incompatible declaration` error if it's outside of the Array class declaration with enable_if. It may be fixed with tricks with template, but keep in it here would be much simpler. So let's keep it here.